exporter/splunkhecexporter/client.go

```diff
@@ -36,25 +37,156 @@ import (
+}
+
+// accept returns true if data is accepted by the buffer
+func (b *bufferState) accept(data []byte) bool {
+	_, err := b.writer.Write(data)
+	overCapacity := errors.Is(err, errOverCapacity)
+	bufLen := b.buf.Len()
+	if overCapacity {
+		bufLen += len(data)
+	}
+	if b.compressionAvailable && !b.compressionEnabled && bufLen > minCompressionLen {
+		// switch over to a zip buffer.
+		tmpBuf := bytes.NewBuffer(make([]byte, 0, b.bufferMaxLen+bufCapPadding))
+		writer := b.gzipWriterPool.Get().(*gzip.Writer)
+		writer.Reset(tmpBuf)
+		zipWriter := &cancellableGzipWriter{
+			innerBuffer: tmpBuf,
+			innerWriter: writer,
+			// 8 bytes required for the zip footer.
+			maxCapacity:    b.bufferMaxLen - 8,
+			gzipWriterPool: b.gzipWriterPool,
+		}
+
+		// the new data is so big, even with a zip writer, we are over the max limit.
+		// abandon and return false, so we can send what is already in our buffer.
+		if _, err2 := zipWriter.Write(b.buf.Bytes()); err2 != nil {
+			return false
+		}
+		b.writer = zipWriter
+		b.buf = tmpBuf
+		b.compressionEnabled = true
+		// if the byte writer was over capacity, try to write the new entry in the zip writer:
+		if overCapacity {
+			if _, err2 := zipWriter.Write(data); err2 != nil {
+				return false
+			}
+
+		}
+		return true
+	}
+	return !overCapacity
+}
+
+type cancellableBytesWriter struct {
+	innerWriter *bytes.Buffer
+	maxCapacity uint
+}
+
+func (c *cancellableBytesWriter) Write(b []byte) (int, error) {
+	if c.maxCapacity == 0 {
+		return c.innerWriter.Write(b)
+	}
+	if c.innerWriter.Len()+len(b) > int(c.maxCapacity) {
+		return 0, errOverCapacity
+	}
+	return c.innerWriter.Write(b)
+}
+
+type cancellableGzipWriter struct {
+	innerBuffer    *bytes.Buffer
+	innerWriter    *gzip.Writer
+	maxCapacity    uint
+	gzipWriterPool *sync.Pool
+	len            int
+}
+
+func (c *cancellableGzipWriter) Write(b []byte) (int, error) {
+	if c.maxCapacity == 0 {
+		return c.innerWriter.Write(b)
+	}
+	c.len += len(b)
+	// if we see that at a 50% compression rate, we'd be over max capacity, start flushing.
+	if (c.len / 2) > int(c.maxCapacity) {
+		// we flush so the length of the underlying buffer is accurate.
+		if err := c.innerWriter.Flush(); err != nil {
+			return 0, err
+		}
+	}
+	// we find that the new content uncompressed, added to our buffer, would overflow our max capacity.
+	if c.innerBuffer.Len()+len(b) > int(c.maxCapacity) {
+		// so we create a copy of our content and add this new data, compressed, to check that it fits.
+		copyBuf := bytes.NewBuffer(make([]byte, 0, c.maxCapacity+bufCapPadding))
+		copyBuf.Write(c.innerBuffer.Bytes())
+		writerCopy := c.gzipWriterPool.Get().(*gzip.Writer)
+		defer c.gzipWriterPool.Put(writerCopy)
+		writerCopy.Reset(copyBuf)
+		if _, err := writerCopy.Write(b); err != nil {
+			return 0, err
+		}
+		if err := writerCopy.Flush(); err != nil {
+			return 0, err
+		}
+		// we find that even compressed, the data overflows.
+		if copyBuf.Len() > int(c.maxCapacity) {
```

Comment:
Shouldn't this be?
```suggestion
		if c.innerBuffer.Len()+copyBuf.Len() > int(c.maxCapacity) {
```

Reply:
No, we have copied c.innerBuffer into copyBuf.

Reply:
Oh right. I missed it